tests/integration/abi_router_test.py

```diff
@@ -0,0 +1,465 @@
+        [
+            (False, OnComplete.ClearStateOC),
+        ],
+        {DRProp.passed: True, DRProp.lastLog: 1},
+    ),
+    (
+        qa_runner("log_creation"),
+        [(True, OnComplete.NoOpOC)],
+        qc_runner("log_creation"),
+        [],
+        {DRProp.passed: True, DRProp.lastLog: "logging creation"},
+    ),
+    (
+        qa_runner(
+            "approve_if_odd"
+        ),  # this should only appear in the clear-state program
+        [],
+        qc_runner("approve_if_odd"),
+        [
+            (False, OnComplete.ClearStateOC),
+        ],
+        {
+            DRProp.passed: lambda args: args[1] % 2 == 1,
+            DRProp.lastLog: None,
+        },
+    ),
+    (
+        qa_runner(None),
+        [(False, OnComplete.OptInOC)],
+        qc_runner(None),
+        [
+            (False, OnComplete.ClearStateOC),
+        ],
+        {
+            DRProp.passed: True,
+            DRProp.lastLog: lambda _, actual: actual
+            in (None, DryRunEncoder.hex("optin call")),
+        },
+    ),
+]
+
+
+def make_yacc_runner(method, approval):
+    return MethodRunner(
+        f"yacc_{'approval' if approval else 'clear'}",
+        approval,
+        YACC_TEAL if approval else YACC_CLEAR_TEAL,
+        CONTRACT,
+        method,
+    )
+
+
+# strip out the bare calls for YACC ~ "yetAnotherContractConstructedFromRouter":
+def yaccify(q_cases):
+    return [
+        (make_yacc_runner(method, True), c[1], make_yacc_runner(method, False), *c[3:])
+        for c in q_cases
+        if (method := c[0].method)
+    ]
+
+
+YACC_CASES = yaccify(QUESTIONABLE_CASES)
+
+ALL_CASES = QUESTIONABLE_CASES + YACC_CASES
+
+
+def get_aa_strat(method_runner, abi_args_mod=None) -> ABIMethodCallStrategy:
+    return ABIMethodCallStrategy(
+        method_runner.contract,
+        method_runner.method,
+        RandomABIStrategyHalfSized,
+        num_dryruns=NUM_ROUTER_DRYRUNS,
+        abi_args_mod=abi_args_mod,
+    )
+
+
+@pytest.mark.parametrize(
+    "approval_runner, approval_call_types, clear_runner, clear_call_types, invariants",
+    ALL_CASES,
+)
+def test_positive(
+    approval_runner, approval_call_types, clear_runner, clear_call_types, invariants
+):
+    """
+    Test the _positive_ version of a case. In other words, ensure that for the given:
+        * method or bare call
+        * OnComplete value
+        * number of arguments
+    that the app call succeeds according to the provided _invariants_ success definition
+    """
+
+    def run_positive(is_approve, method_runner, call_types, invariants):
+        good_abi_args = get_aa_strat(method_runner)
+
+        algod = get_algod()
+
+        if not call_types:
+            return
+
+        method = method_runner.method
+        sim = Simulation(
+            algod,
+            ExecutionMode.Application,
+            method_runner.teal,
+            invariants,
+            abi_method_signature=good_abi_args.method_signature(),
+        )
```

Comment:
## STEP 1. Prepare a `Simulation` object for executing dry runs, and asserting that it behaved as expected
It needs:
* an `algod`
* a `mode`
* `teal` program
* `invariants` which is a "predicates" dictionary mapping from `DryRunProperty`s to expressions that should evalute true for a dry run execution
* an `abi_method_signature` which will be used for validating inputs and decoding outputs